lib/fog/openstack/core.rb

```diff
@@ -504,17 +503,19 @@ def self.retrieve_tokens_v3(options, connection_options = {})
       end
       request_body[:auth][:scope] = scope unless scope.empty?
 
-      puts "request_body: #{request_body}" if user_domain=='Default2'
+      path     = (uri.path and not uri.path.empty?) ? uri.path : 'v3'
 
-      response = connection.request({
-                                        :expects => [201],
-                                        :headers => {'Content-Type' => 'application/json'},
-                                        :body => Fog::JSON.encode(request_body),
-                                        :method => 'POST',
-                                        :path => (uri.path and not uri.path.empty?) ? uri.path : 'v2.0'
-                                    })
+      response, expires = @@token_cache[{body: request_body, path: path}]
 
-      puts "response.body: #{response.body}" if user_domain=='Default2'
+      unless response && expires > Time.now
```

Comment:
Not sure I agree - the logic here is: "if there is a cached response, and if the expiry time is in the future (i.e. greater than 'now'), then don't bother making a request".